### ISSUE TITLE
Appointed status double and Double invoice

### ISSUE THREAD

**Issue by omargits** [PAYONE-GmbH/magento-2#296]
@niclashoyer  @fjbender @hreinberger @FatchipRobert
When order placed by Payone Paypal always two invoices created one is with original amount and other for 0.00 amount also we having two Appointed Status.

This is start happening sudden from last week. where as in last two weeks we didn't change any thing in db or files. Server is very normal no load on server. very less users and not a heavy catalog

Please guide how to fixed it and also if it from magento side how to catch this issue for further debugging 

> - 2***:f**:**08::***:*4*:1* - - [09/Mar/2020:23:09:15 +0100] "POST /payone/transactionstatus/decouple/ HTTP/1.1" 302 4268 "-" "-"
>- 1*5.*0.*0.* - - [09/Mar/2020:23:09:14 +0100] "POST /payone/transactionstatus HTTP/1.1" 200 4093 "-" "PAYONE FinanceGate"

**Comment by fjbender** [comment]
Hi @omargits,

this looks weird and is most likely a configuration issue. Please contact the Technical Support stating your MID and any TXID of affected transactions. Please also make sure that in the portal configuration (PMI) the configured Transaction Status URL is correct.

**Comment by omargits** [comment]
@niclashoyer @fjbender @hreinberger @FatchipRobert we cross checked all are fine. Even server support double check the server specially at transaction time server resources are fully free also with not customers burden on server.

a week ago its was working fine. after that there is no change. We already write to support also they said plz contact the server. server say contact you guys kindly guide me

**Comment by fjbender** [comment]
Can you provide me with a TXID of one affected transaction? I'd need to see exactly what is happening on our side.

**Comment by omargits** [comment]
do i email u or is any way to DM you ?

**Comment by fjbender** [comment]
You can mail it to integrations@payone.com

**Comment by omargits** [comment]
email send plz check and guide us dear.

**Comment by fjbender** [comment]
Okay, that helped me a lot, thanks. For each of the transactions you listed, we have logged one retry, meaning that we resent the notification. However, I can't see what was the failed cause of the first try. I'll ask if we have that logged elsewhere.

**Comment by omargits** [comment]
@fjbender Thanks dear will be really great if u help us with detail in this matter and try to solve this. this is start happening last week before it was working fine. Your efforts in this will be highly appreciated

**Comment by fjbender** [comment]
Unfortunately, we do not retain previous responses, so I'd need more logs from your side. You'd need to add some logging to the Transaction Status processor. Ideally, you can tell me the runtime and response of transaction status queries from our platform to your shop. Our platform will retry the transaction status if either
a) the runtime exceeds 10 seconds
b) the response does not start with the string "TSOK"

**Comment by omargits** [comment]
@fjbender 
I reply via email there i send u some transaction detail log plz have a look kind have a look there

**Comment by fjbender** [comment]
Thanks, so we can probably rule out the timeout option. Probably there is an invalid response to the first status while it still gets processed. From the logs you sent me I cannot see the actual response that was sent to our platform. Do you have a possibility to log the actual response?

Typical reasons for this type of error are
- Answering with a HTTP 302 instead of 200 (our platform does not follow 302s; however your logs indicate otherwise)
- PHP errors or warnings that are sent before the TSOK string (not common in production environments though)
- Weird load balancer/reverse proxy behavior

**Comment by omargits** [comment]
can you guide us little how to catch and that u type of log u looking for how we can reproduce it.

**Comment by fjbender** [comment]
I'd start by logging any of the return statements in `\Payone\Core\Controller\Transactionstatus\Index::handleTransactionStatus` with the respective `txid` obtained through `$this->getParam('txid');`.

If that didn't help, we'd need to go deeper to make sure the webserver doesn't write any messages before this controller does its job.

**Comment by andreasberenz** [comment]
We're facing the same (or a similar) issue here.
Transactions seemingly have timed out, but i monitored our server here and made the following observations:

there can be a gap between the preauthorization and the transactionstatus. In one example we have the preauthorization at 16:02. At 16:05 the request with the tansactionstatus came in. Both timestamps are logged in magento, as they occured.
The status-mapping worked fine as well and the request was responded with the status 200.
But if I have a look at the merchant center, I can find this transactionstatus open with one try and a reported timeout.
The creation timestamp is here 16:02 but the earliest retry is 17:05 - exactly one hour after the transactionstatus-request came in. If these are atomic requests here you obviously have the right request minute here, otherwise I can't explain how this gap could arise.
I double checked other transactions and the retry time is always exactly one hour later than the time, the request arrived our server. 

Could you please check on your server side if everything is correct with your notification-service?
We do not create invoices directly if the payment is successful, but we have another issue here, that magento creates new orders, when the original order was replaced and the transactionstatus request is processed.

Surely this  misbehaviour is not only a payone issue, but these timeouts definetely did not occur in the time I monitored this. I did not log the resonse our server sent, maybe there is a problem as well in some edge cases.

**Comment by fjbender** [comment]
We use this retry mechanism starting with a waiting period of one hour whenever we do not receive the response we need (a plain string "TSOK") within a timeout of 10 seconds. There are various reasons for this happening that we have observed in the wild. Most common issues include some rogue transaction status forwarding (though we tried to harden this in the latest versions), a database lock, or a reverse proxy doing something weird.

If you send over the transaction ID (txid) of one or two affected transaction to integrations@payone.com, I'll try to find out more. However, logs of failed transaction status delivery attempts are not always retained, so writing a log of the actual response on your side might be helpful too.
